### ISSUE TITLE
implicit Deref coercion not happening

### ISSUE THREAD

**Issue by rbtcollins** [rust-lang/rust#109829]
```rust
use std::path::{Path,PathBuf};

#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord)]
pub(crate) struct PathBasedToolchain(PathBuf);

impl TryFrom<&Path> for PathBasedToolchain {
    type Error = anyhow::Error;

    fn try_from(value: &Path) -> std::result::Result<Self, Self::Error> {
                Ok(PathBasedToolchain(value.into()))
    }
}


fn main() -> anyhow::Result<()> {
    let path = PathBuf::new();
    let t =  PathBasedToolchain::try_from(&path);
    let u =  PathBasedToolchain::try_from(&path as &Path);
    
    Ok(())
}
```
(playground link)[https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=53696b2684233ffd81680129f6c076d2]

I expected to see this happen: compile with no error (because of https://doc.rust-lang.org/book/ch15-02-deref.html#implicit-deref-coercions-with-functions-and-methods and https://doc.rust-lang.org/stable/std/path/struct.PathBuf.html#impl-Deref-for-PathBuf ).

Instead, this happened:

```
Compiling playground v0.0.1 (/playground)
error[[E0277]](https://doc.rust-lang.org/nightly/error_codes/E0277.html): the trait bound `PathBasedToolchain: From<&PathBuf>` is not satisfied
  --> src/lib.rs:17:43
   |
17 |     let t =  PathBasedToolchain::try_from(&path);
   |              ---------------------------- ^^^^^ the trait `From<&PathBuf>` is not implemented for `PathBasedToolchain`
   |              |
   |              required by a bound introduced by this call
   |
   = help: the trait `TryFrom<&Path>` is implemented for `PathBasedToolchain`
   = note: required for `&PathBuf` to implement `Into<PathBasedToolchain>`
   = note: required for `PathBasedToolchain` to implement `TryFrom<&PathBuf>`

error[[E0277]](https://doc.rust-lang.org/nightly/error_codes/E0277.html): the trait bound `PathBasedToolchain: From<&PathBuf>` is not satisfied
  --> src/lib.rs:17:14
   |
17 |     let t =  PathBasedToolchain::try_from(&path);
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `From<&PathBuf>` is not implemented for `PathBasedToolchain`
   |
   = help: the trait `TryFrom<&Path>` is implemented for `PathBasedToolchain`
   = note: required for `&PathBuf` to implement `Into<PathBasedToolchain>`
   = note: required for `PathBasedToolchain` to implement `TryFrom<&PathBuf>`
```

### Meta

`rustc --version --verbose`:
```
Nightly channel

Build using the Nightly version: 1.70.0-nightly

(2023-03-31 5e1d3299a290026b8578)
```

This doesn't cause a backtrace from the compiler.
<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

**Comment by SkiFire13** [comment]
AFAIK this is expected, the compiler [doesn't do coercions when matching traits](https://doc.rust-lang.org/nomicon/coercions.html) (third paragraph), i.e. here the compiler is looking for an implementation of `TryFrom<typeof(&path)>` for `PathBasedToolchain` (the syntax `typeof(&path)` doesn't exist but I hope the meaning is clear), but that doesn't exist, and the implementation of `TryFrom<&Path>` does not constitute an implementation of `TryFrom<&PathBuf>` too.

**Comment by rbtcollins** [comment]
Thank you. It might be nice to link to the nomicon from the docs on implicit coercions, because while the nomicon is clear, I didn't find it after non-trivial googling.

**Comment by chenyukang** [comment]
But why there are two almost the same diagnostics show up?
I think the second one could be removed.

**Comment by Spartan2909** [comment]
I've opened https://github.com/rust-lang/book/pull/3670 to try and clarify this behaviour.

**Comment by oli-obk** [comment]
Unassigning due to inactivity

**Comment by ArtemIsmagilov** [comment]
There seems to be a missing help message here about using the `as` operator to get the code to compile.

help: try add `as` operator: PathBasedToolchain::try_from(&path as &Path);
_______________________________________________________________+++++++

**Comment by PNDTG** [comment]
I think this issues needs to get closed. It either needs to explicitly get PathBuf and then deref to get Path in the trait function or be given Path.
In the documents it is explicitly stated that "If Rust didn’t implement deref coercion, we would have to write the code in Listing 15-13 instead of the code in Listing 15-12 to call hello with a value of type &MyBox<String>." [Docs listing](https://doc.rust-lang.org/book/ch15-02-deref.html#listing-15-12) it's obviously that Rust needs to have implemented that coercion type(or implicitly derived from the rules.)

Maybe we could have it be boldened and or give a small reference to allowed coercions. [Coercion types](https://doc.rust-lang.org/reference/type-coercions.html#coercion-types)
I could do that...
